### ISSUE TITLE
cython master fails to cythonize pyFAI

### ISSUE THREAD

**Issue by tacaswell** [cython/cython#3560]
Trying to build https://github.com/silx-kit/pyFAI I am getting exceptions on cython master.  bisecting says 7d99b0f085bb254e9a9b4a9fe50aefe133837a81 is the first bad commit

attn @kif 

In particular https://github.com/silx-kit/pyFAI/blob/ed24539306aeb6834db240ce98abd8827be609ab/pyFAI/ext/preproc.pyx fails with

```
jupiter@14:47  ➤  cythonize  pyFAI/ext/preproc.pyx
Compiling /home/tcaswell/source/bnl/esrf/pyFAI/pyFAI/ext/preproc.pyx because it changed.
[1/1] Cythonizing /home/tcaswell/source/bnl/esrf/pyFAI/pyFAI/ext/preproc.pyx
/home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/Main.py:344: FutureWarning: Cython directive 'language_level' not set, using '3str' for now (Py3). This has changed from earlier releases! File: /home/tcaswell/source/bnl/esrf/pyFAI/pyFAI/ext/preproc.pyx
  tree = Parsing.p_module(s, pxd, full_module_name)
Traceback (most recent call last):
  File "/home/tcaswell/.virtualenvs/bleeding/bin/cythonize", line 8, in <module>
    sys.exit(main())
  File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Build/Cythonize.py", line 224, in main
    cython_compile(path, options)
  File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Build/Cythonize.py", line 68, in cython_compile
    ext_modules = cythonize(
  File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Build/Dependencies.py", line 1105, in cythonize
    cythonize_one(*args)
  File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Build/Dependencies.py", line 1246, in cythonize_one
    result = compile_single(pyx_file, options, full_module_name=full_module_name)
  File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/Main.py", line 575, in compile_single
    return run_pipeline(source, options, full_module_name)
  File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/Main.py", line 503, in run_pipeline
    err, enddata = Pipeline.run_pipeline(pipeline, source)
  File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/Pipeline.py", line 356, in run_pipeline
    data = run(phase, data)
  File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/Pipeline.py", line 336, in run
    return phase(data)
  File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/Pipeline.py", line 52, in generate_pyx_code_stage
    module_node.process_implementation(options, result)
  File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/ModuleNode.py", line 161, in process_implementation
    self.generate_c_code(env, options, result)
  File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/ModuleNode.py", line 412, in generate_c_code
    self.body.generate_function_definitions(env, code)
  File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/Nodes.py", line 398, in generate_function_definitions
    stat.generate_function_definitions(env, code)
  File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/Nodes.py", line 398, in generate_function_definitions
    stat.generate_function_definitions(env, code)
  File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/Nodes.py", line 345, in generate_function_definitions
    self.body.generate_function_definitions(env, code)
  File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/Nodes.py", line 398, in generate_function_definitions
    stat.generate_function_definitions(env, code)
  File "Cython/Compiler/FusedNode.py", line 899, in Cython.Compiler.FusedNode.FusedCFuncDefNode.generate_function_definitions
  File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/Nodes.py", line 2083, in generate_function_definitions
    assert gil_owned['error'] == gil_owned['success'], "%s != %s" % (gil_owned['error'], gil_owned['success'])
AssertionError: True != False
```

Sorry I don't have a better reproduction case.

----

output from build

<details>

```
< --- snip --- >
    cythoning pyFAI/ext/splitPixelFullCSR.pyx to pyFAI/ext/splitPixelFullCSR.c
    /home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/Main.py:344: FutureWarning: Cython directive 'language_level' not set, using '3str' for now (Py3). This has changed from earlier releases! File: /tmp/pip-req-build-frrjpdam/pyFAI/ext/splitPixelFullCSR.pyx
      tree = Parsing.p_module(s, pxd, full_module_name)
    cythoning pyFAI/ext/relabel.pyx to pyFAI/ext/relabel.c
    /home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/Main.py:344: FutureWarning: Cython directive 'language_level' not set, using '3str' for now (Py3). This has changed from earlier releases! File: /tmp/pip-req-build-frrjpdam/pyFAI/ext/relabel.pyx
      tree = Parsing.p_module(s, pxd, full_module_name)
    cythoning pyFAI/ext/bilinear.pyx to pyFAI/ext/bilinear.c
    /home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/Main.py:344: FutureWarning: Cython directive 'language_level' not set, using '3str' for now (Py3). This has changed from earlier releases! File: /tmp/pip-req-build-frrjpdam/pyFAI/ext/bilinear.pyx
      tree = Parsing.p_module(s, pxd, full_module_name)
    cythoning pyFAI/ext/_bispev.pyx to pyFAI/ext/_bispev.c
    /home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/Main.py:344: FutureWarning: Cython directive 'language_level' not set, using '3str' for now (Py3). This has changed from earlier releases! File: /tmp/pip-req-build-frrjpdam/pyFAI/ext/_bispev.pyx
      tree = Parsing.p_module(s, pxd, full_module_name)
    cythoning pyFAI/ext/_convolution.pyx to pyFAI/ext/_convolution.c
    /home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/Main.py:344: FutureWarning: Cython directive 'language_level' not set, using '3str' for now (Py3). This has changed from earlier releases! File: /tmp/pip-req-build-frrjpdam/pyFAI/ext/_convolution.pyx
      tree = Parsing.p_module(s, pxd, full_module_name)
    cythoning pyFAI/ext/_blob.pyx to pyFAI/ext/_blob.c
    /home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/Main.py:344: FutureWarning: Cython directive 'language_level' not set, using '3str' for now (Py3). This has changed from earlier releases! File: /tmp/pip-req-build-frrjpdam/pyFAI/ext/_blob.pyx
      tree = Parsing.p_module(s, pxd, full_module_name)
    cythoning pyFAI/ext/morphology.pyx to pyFAI/ext/morphology.c
    /home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/Main.py:344: FutureWarning: Cython directive 'language_level' not set, using '3str' for now (Py3). This has changed from earlier releases! File: /tmp/pip-req-build-frrjpdam/pyFAI/ext/morphology.pyx
      tree = Parsing.p_module(s, pxd, full_module_name)
    cythoning pyFAI/ext/watershed.pyx to pyFAI/ext/watershed.c
    /home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/Main.py:344: FutureWarning: Cython directive 'language_level' not set, using '3str' for now (Py3). This has changed from earlier releases! File: /tmp/pip-req-build-frrjpdam/pyFAI/ext/watershed.pyx
      tree = Parsing.p_module(s, pxd, full_module_name)
    cythoning pyFAI/ext/_tree.pyx to pyFAI/ext/_tree.c
    /home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/Main.py:344: FutureWarning: Cython directive 'language_level' not set, using '3str' for now (Py3). This has changed from earlier releases! File: /tmp/pip-req-build-frrjpdam/pyFAI/ext/_tree.pyx
      tree = Parsing.p_module(s, pxd, full_module_name)
    cythoning pyFAI/ext/sparse_utils.pyx to pyFAI/ext/sparse_utils.c
    /home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/Main.py:344: FutureWarning: Cython directive 'language_level' not set, using '3str' for now (Py3). This has changed from earlier releases! File: /tmp/pip-req-build-frrjpdam/pyFAI/ext/sparse_utils.pyx
      tree = Parsing.p_module(s, pxd, full_module_name)
    cythoning pyFAI/ext/preproc.pyx to pyFAI/ext/preproc.c
    /home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/Main.py:344: FutureWarning: Cython directive 'language_level' not set, using '3str' for now (Py3). This has changed from earlier releases! File: /tmp/pip-req-build-frrjpdam/pyFAI/ext/preproc.pyx
      tree = Parsing.p_module(s, pxd, full_module_name)
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-req-build-frrjpdam/setup.py", line 1108, in <module>
        setup_package()
      File "/tmp/pip-req-build-frrjpdam/setup.py", line 1104, in setup_package
        setup(**setup_kwargs)
      File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/setuptools/__init__.py", line 144, in setup
        return distutils.core.setup(**attrs)
      File "/home/tcaswell/.pybuild/bleeding/lib/python3.9/distutils/core.py", line 148, in setup
        dist.run_commands()
      File "/home/tcaswell/.pybuild/bleeding/lib/python3.9/distutils/dist.py", line 966, in run_commands
        self.run_command(cmd)
      File "/home/tcaswell/.pybuild/bleeding/lib/python3.9/distutils/dist.py", line 985, in run_command
        cmd_obj.run()
      File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/setuptools/command/install.py", line 61, in run
        return orig.install.run(self)
      File "/home/tcaswell/.pybuild/bleeding/lib/python3.9/distutils/command/install.py", line 546, in run
        self.run_command('build')
      File "/home/tcaswell/.pybuild/bleeding/lib/python3.9/distutils/cmd.py", line 313, in run_command
        self.distribution.run_command(command)
      File "/home/tcaswell/.pybuild/bleeding/lib/python3.9/distutils/dist.py", line 985, in run_command
        cmd_obj.run()
      File "/home/tcaswell/.pybuild/bleeding/lib/python3.9/distutils/command/build.py", line 135, in run
        self.run_command(cmd_name)
      File "/home/tcaswell/.pybuild/bleeding/lib/python3.9/distutils/cmd.py", line 313, in run_command
        self.distribution.run_command(command)
      File "/home/tcaswell/.pybuild/bleeding/lib/python3.9/distutils/dist.py", line 984, in run_command
        cmd_obj.ensure_finalized()
      File "/home/tcaswell/.pybuild/bleeding/lib/python3.9/distutils/cmd.py", line 107, in ensure_finalized
        self.finalize_options()
      File "/tmp/pip-req-build-frrjpdam/setup.py", line 588, in finalize_options
        build_ext.finalize_options(self)
      File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/setuptools/command/build_ext.py", line 144, in finalize_options
        self.check_extensions_list(self.extensions)
      File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Distutils/old_build_ext.py", line 192, in check_extensions_list
        ext.sources = self.cython_sources(ext.sources, ext)
      File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Distutils/old_build_ext.py", line 342, in cython_sources
        result = cython_compile(source, options=options,
      File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/Main.py", line 626, in compile
        return compile_single(source, options, full_module_name)
      File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/Main.py", line 575, in compile_single
        return run_pipeline(source, options, full_module_name)
      File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/Main.py", line 503, in run_pipeline
        err, enddata = Pipeline.run_pipeline(pipeline, source)
      File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/Pipeline.py", line 356, in run_pipeline
        data = run(phase, data)
      File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/Pipeline.py", line 336, in run
        return phase(data)
      File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/Pipeline.py", line 52, in generate_pyx_code_stage
        module_node.process_implementation(options, result)
      File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/ModuleNode.py", line 161, in process_implementation
        self.generate_c_code(env, options, result)
      File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/ModuleNode.py", line 412, in generate_c_code
        self.body.generate_function_definitions(env, code)
      File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/Nodes.py", line 398, in generate_function_definitions
        stat.generate_function_definitions(env, code)
      File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/Nodes.py", line 398, in generate_function_definitions
        stat.generate_function_definitions(env, code)
      File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/Nodes.py", line 345, in generate_function_definitions
        self.body.generate_function_definitions(env, code)
      File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/Nodes.py", line 398, in generate_function_definitions
        stat.generate_function_definitions(env, code)
      File "Cython/Compiler/FusedNode.py", line 899, in Cython.Compiler.FusedNode.FusedCFuncDefNode.generate_function_definitions
      File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.9/site-packages/Cython/Compiler/Nodes.py", line 2081, in generate_function_definitions
        assert gil_owned['error'] == gil_owned['success'], "%s != %s" % (gil_owned['error'], gil_owned['success'])
    AssertionError: True != False
    ----------------------------------------
```

</details>

git bisect log

<details>

I started this when 01085d2ad588bf41e2d17cff6aa96d0507087ce3 was master, but have verified that it does not work with 792d80bade7138993ec3b1d2fb3133dced2bc871 either.

```sh
jupiter@14:34  ➤  git bisect log
git bisect start
# good: [203ec86c39165c4001aa9238268f1911db1e8728] Update and split changelog for 3.0a2.
git bisect good 203ec86c39165c4001aa9238268f1911db1e8728
# bad: [01085d2ad588bf41e2d17cff6aa96d0507087ce3] Make the GIL-avoidance in 7d99b0f0 actually work in nogil functions and not just nogil sections. Closes #3558.
git bisect bad 01085d2ad588bf41e2d17cff6aa96d0507087ce3
# good: [4dbce65d359473c1dc37eb0c325ae177aa0774a2] Clean up and repair the type spec generation for the limited API, leaving more of the work to the TypeSlots instead of special-casing all sorts of things.
git bisect good 4dbce65d359473c1dc37eb0c325ae177aa0774a2
# good: [cf5e3fe3520f51cdf001956e81bd4fdb5e86233b] Fix reference to missing utility function, which is not more than an alias to CPython's "PyObject_GenericGetAttr" in the limited API case.
git bisect good cf5e3fe3520f51cdf001956e81bd4fdb5e86233b
# bad: [89172b6d6d40ef0fea20e6b0c37aab2148daedd3] Add a test for inferred integer %-operations.
git bisect bad 89172b6d6d40ef0fea20e6b0c37aab2148daedd3
# good: [25b7d7e415d179077a6d8ee150d9bc59bc5e09e0] Relax class private names for c types (GH-3546)
git bisect good 25b7d7e415d179077a6d8ee150d9bc59bc5e09e0
# bad: [a1a2bd70c66c8905d4011d189ac5fd8d38134521] Update changelog.
git bisect bad a1a2bd70c66c8905d4011d189ac5fd8d38134521
# bad: [7d99b0f085bb254e9a9b4a9fe50aefe133837a81] Avoid acquiring the GIL at the end of nogil functions (GH-3556)
git bisect bad 7d99b0f085bb254e9a9b4a9fe50aefe133837a81
# first bad commit: [7d99b0f085bb254e9a9b4a9fe50aefe133837a81] Avoid acquiring the GIL at the end of nogil functions (GH-3556)
(bleeding) ✔  ~/source/other_source/cython [:7d99b0f08|BISECTING|⚑ 2]
```

</details>

**Comment by scoder** [comment]
Thanks for the report. The assertion actually wasn't correct – but it was good to have it, because it also pointed at a few real bugs. Turns out that the error exit path of a nogil function is really more likely to require the GIL for cleanup actions than the success path – which is a good thing. So I chose to degrade the error case in favour of less GIL handling on the success path in dc41690ac13a25253ac3790cb112d557a2430aae.

I tried it with your code and it compiled for me. But it's also worth running some actual tests over it, I think.
